### PR TITLE
Add new rule no_catch_expressions

### LINKED ISSUE #145: New Rule: Avoid catch expressions
I'm specifically talking about `catch 1+a` stuff, like we find here: http://erlang.org/doc/reference_manual/expressions.html#catch.

This'll generate an unused stacktrace as well as "incomprehensible" return, as explained in

```
For exceptions of class error, that is, run-time errors, {'EXIT',{Reason,Stack}} is returned.

For exceptions of class exit, that is, the code called exit(Term), {'EXIT',Term} is returned.

For exceptions of class throw, that is the code called throw(Term), Term is returned.
```

@elbrujohalcon: thoughts?

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
Closes #145.

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
The Wiki was also updated, having as target 1.4.0. (most surely it'll be 2.0.0, but we can adapt later)

**Comment by paulo-ferraz-oliveira** [comment]
I'm finding some patterns in the approaches to coding new rules, as time goes by. I'll probably to some refactoring at a give moment, to:
1. increase code readability,
2. ease integration of new changes (especially by new contributors).

File: test/examples/fail_invalid_dynamic_call.erl [hunk 1]
```diff
@@ -3,7 +3,7 @@
 -ignore_xref({normal, call, 0}).
 -ignore_xref({another_normal, call, 0}).
 
--dialyzer(no_match).
+-dialyzer(no_match). -elvis([{elvis_style, no_catch_expressions, disable}]).
```
Comment (elbrujohalcon) [inline test/examples/fail_invalid_dynamic_call.erl:6]: What happened here? Why are these two attributes on the same line?
```suggestion
-dialyzer(no_match).
-elvis([{elvis_style, no_catch_expressions, disable}]).
```

File: test/examples/pass_invalid_dynamic_call_elvis_attr.erl [hunk 2]
```diff
@@ -4,7 +4,7 @@
 -elvis([{elvis_text_style, line_length, #{limit => 100}}]).
 -elvis([{elvis_style, atom_naming_convention, #{regex => "^([a-z][a-z0-9]*_?_?)*(_SUITE)?$"}}]).
 
--dialyzer(no_match).
+-dialyzer(no_match). -elvis([{elvis_style, no_catch_expressions, disable}]).
```
Comment (elbrujohalcon) [inline test/examples/pass_invalid_dynamic_call_elvis_attr.erl:7]:
```suggestion
-dialyzer(no_match).
-elvis([{elvis_style, no_catch_expressions, disable}]).
```

File: test/examples/pass_nesting_level_elvis_attr.erl [hunk 3]
```diff
@@ -3,7 +3,7 @@
 -elvis([{elvis_style, nesting_level, #{level => 5}}, {elvis_style, no_if_expression, disable}]).
 -elvis([{elvis_text_style, line_length, #{limit => 100}}]).
 
--dialyzer(no_match).
+-dialyzer(no_match). -elvis([{elvis_style, no_catch_expressions, disable}]).
```
Comment (elbrujohalcon) [inline test/examples/pass_nesting_level_elvis_attr.erl:6]:
```suggestion
-dialyzer(no_match).
-elvis([{elvis_style, no_catch_expressions, disable}]).
```

**Review by elbrujohalcon — APPROVED** [review]

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
Oops! I meant to request changes, not approve.

File: test/examples/fail_invalid_dynamic_call.erl (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply test/examples/fail_invalid_dynamic_call.erl:6]: I didn't want to have to update the tests that rely on line numbers.
Do you prefer them on separate lines?

File: test/examples/fail_no_catch_expressions.erl [hunk 4]
```diff
@@ -0,0 +1,25 @@
+-module(fail_no_catch_expressions).
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply test/examples/fail_no_catch_expressions.erl:1]: No at all.

File: test/examples/pass_invalid_dynamic_call_elvis_attr.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/examples/pass_invalid_dynamic_call_elvis_attr.erl:7]: Same reasoning as above. Do I change all the tests where the lines got shifted?

File: test/examples/pass_nesting_level_elvis_attr.erl (hunk 3 above)
Comment (paulo-ferraz-oliveira) [reply test/examples/pass_nesting_level_elvis_attr.erl:6]: 😄

**Comment by paulo-ferraz-oliveira** [comment]
I've added more "catch" stuff to the test/example file. Let me know.

File: test/examples/fail_invalid_dynamic_call.erl (hunk 1 above)
Comment (elbrujohalcon) [reply test/examples/fail_invalid_dynamic_call.erl:6]: There is an empty line above. Let's use that one.

File: test/examples/pass_invalid_dynamic_call_elvis_attr.erl (hunk 2 above)
Comment (paulo-ferraz-oliveira) [reply test/examples/pass_invalid_dynamic_call_elvis_attr.erl:7]: 👍

File: test/examples/pass_nesting_level_elvis_attr.erl (hunk 3 above)
Comment (paulo-ferraz-oliveira) [reply test/examples/pass_nesting_level_elvis_attr.erl:6]: 👍

**Pull request merged**